### ISSUE TITLE
Handle foreign keys to dropped child tables

### ISSUE THREAD

**Issue by keithf4** [pgpartman/pg_partman#584]
MAINTAINER NOTE: Creating a new issue to discuss/track this

I am still extending my partitioning process and I found a new issue more or less related to this ticket. 

When I detach my partitions (without dropping them) in this situation, it works but breaks my foreign keys because they still target the main table but the partitions are no longer present. I think I will call a function after detaching to remove all foreign keys from detached partitions.

If you want to handle this issue in a future version, it might be good to (optionally?) drop foreign keys on detached partitions.

_Originally posted by @axelfaure in https://github.com/pgpartman/pg_partman/issues/497#issuecomment-1363012216_

**Comment by keithf4** [comment]
So just to clarify, you had foreign keys coming INTO the partition set that were broken when child tables in that referenced partition set were removed? Or were the foreign keys FROM the partition set to another table?

**Comment by axelfaure** [comment]
In my case it was actually both. I was in a situation where table A referenced B and B referenced C (and so on). 
A, B and C were all partitioned by date an I periodically created new partitions and detached old ones.

In this case, I started by triggering `run_maintenance` on A (no problem at that point I ended up with A still referencing B and the now detached partitions from A also referencing B).
Then I triggered `run_maintenance` on B and that is were things starting to be a bit more ugly. A was still referencing B and since the partitioning scheme was the same on both tables that part was coherent. But detached partitions from A also referenced B and that is the issue. The data actually referenced by these partitions was now in the detached partitions from B and no longer in B so the referential integrity was broken. 
And as I kept running `run_maintenance` on more partitions the problem was becoming worse and worse.

I don't think there is actually an easy "catch all" solution to fully solve this problem since the issue mostly comes from the fact that I had FK coming INTO the maintained table and I don't think pg_partman should change these. 
However, I also think that statically speaking this partitioning scheme (partitioning several tables using the same key so that partition 1 in A actually references partition 1 in B) is not uncommon. Even more so when we focus on systems where partitions need to be automatically created/removed. 
As a result I think offering a way to automatically remove all FK from the detached partitions could help many users of pg_partman keeping their db structure coherent.

**Comment by axelfaure** [comment]
In my case the FK were declared from a parent table to a parent table. Creating FK between specific partitions might be a solution to this issue but it does not seem that easy to automate.

The solution I found to my issue was to detach partition in the correct order (in my example first A, then B, then C) and to drop all FKs on the newly detached partition before detaching the next one. 

That is not a good generic solution since it only work if :
1. there are no other references to this data in unpartitioned tables
2. there is no cyclic reference (in which case there wouldn't be a correct order) 

But I think (and I might very well be wrong) that it could work 80% of the time on real world systems. Combined with an integrity check preventing the user to break its data for the remaining 20% it could still be an improvement.

In terms of new features, pg_partman would only need the ability to perform maintenance in a specified order (which you already implemented I think) and a flag that, if set, would drop the FK when detaching a partition (plus the integrity check if you want to add it).

**Comment by keithf4** [comment]
I do have the fixed maintenance order committed for the next release. 
I'll have to see about the FK option and how well it works in a general sense.
Thanks for reporting the issues!

**Comment by keithf4** [comment]
Version 5.5.0 has been released with a configurable flag of whether to detach the child table before dropping it. Couldn't find an easy universal solution to automatically detect whether to detach first or not, so just made it a user toggle for now so people can hopefully move forward for that partition set's particular needs.

Sorry for the delay. Let me know if it's working ok for you.

**Comment by axelfaure** [comment]
Hi ! Thanks for the update. 

I am not entirely sure I understand in what exact situation this flag is meant to be used. Reading its description in the changelog, I don't think it addresses the issue I was mentioning back then since I was not dropping the partitions, only detaching them. What needed to be dropped were the foreign keys on the detached partitions (I think).

To be honest the system in question has evolved quite a bit in the past 2 or 3 years so I probably won't be able to get fresh updates on how well this use case is handled by partman now. But as I recall it (and from what I get from our messages), I think it is a different issue.

**Comment by keithf4** [comment]
You're right this doesn't fully apply in this case. Had a lot of issues open around foreign keys, so I was mostly replying to all of them to let them know of the included update.

I'll come back to revisit this one again in the future.
